File: Aura-Frontend/Aura/src/services/followService.js
import api from './api';

const followService = {
    // Follow a user
    follow: async (userId) => {
        try {
            const response = await api.post(`/follow/user/${userId}`);
            return {
                success: true,
                message: response.data.message || 'Successfully followed user'
            };
        } catch (error) {
            console.error('Follow failed:', error);
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to follow user'
            };
        }
    },

    // Unfollow a user
    unfollow: async (userId) => {
        try {
            const response = await api.delete(`/follow/user/${userId}`);
            return {
                success: true,
                message: response.data.message || 'Successfully unfollowed user'
            };
        } catch (error) {
            console.error('Unfollow failed:', error);
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to unfollow user'
            };
        }
    },

    // Check if following a user
    isFollowing: async (userId) => {
        try {
            const response = await api.get(`/follow/check/${userId}`);
            return {
                success: true,
                isFollowing: response.data.isFollowing || false
            };
        } catch (error) {
            console.error('Check following failed:', error);
            return {
                success: false,
                isFollowing: false,
                error: error.response?.data?.message || 'Failed to check follow status'
            };
        }
    },

    // Get list of followers for a user
    getFollowers: async (userId, page = 0, size = 20) => {
        try {
            const response = await api.get(`/follow/followers/${userId}`);
            return {
                success: true,
                followers: response.data.content || response.data || [],
                totalElements: response.data.totalElements,
                totalPages: response.data.totalPages
            };
        } catch (error) {
            console.error('Get followers failed:', error);
            return {
                success: false,
                followers: [],
                error: error.response?.data?.message || 'Failed to fetch followers'
            };
        }
    },

    // Get list of users that a user is following
    getFollowing: async (userId, page = 0, size = 20) => {
        try {
            const response = await api.get(`/follow/following/${userId}`);
            return {
                success: true,
                following: response.data.content || response.data || [],
                totalElements: response.data.totalElements,
                totalPages: response.data.totalPages
            };
        } catch (error) {
            console.error('Get following failed:', error);
            return {
                success: false,
                following: [],
                error: error.response?.data?.message || 'Failed to fetch following'
            };
        }
    },

    // Get follower/following stats for a user
    getFollowStats: async (userId) => {
        try {
            const response = await api.get(`/follow/stats/${userId}`);
            return {
                success: true,
                stats: {
                    followerCount: response.data.followerCount || 0,
                    followingCount: response.data.followingCount || 0
                }
            };
        } catch (error) {
            console.error('Get follow stats failed:', error);
            return {
                success: false,
                stats: {
                    followerCount: 0,
                    followingCount: 0
                },
                error: error.response?.data?.message || 'Failed to fetch follow stats'
            };
        }
    },

    // Follow an artist
    followArtist: async (artistId) => {
        try {
            const response = await api.post(`/follow/artist/${artistId}`);
            return {
                success: true,
                message: response.data.message || 'Successfully followed artist'
            };
        } catch (error) {
            console.error('Follow artist failed:', error);
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to follow artist'
            };
        }
    },

    // Unfollow an artist
    unfollowArtist: async (artistId) => {
        try {
            const response = await api.delete(`/follow/artist/${artistId}`);
            return {
                success: true,
                message: response.data.message || 'Successfully unfollowed artist'
            };
        } catch (error) {
            console.error('Unfollow artist failed:', error);
            return {
                success: false,
                error: error.response?.data?.message || 'Failed to unfollow artist'
            };
        }
    },

    // Check if following an artist
    isFollowingArtist: async (artistId) => {
        try {
            const response = await api.get(`/follow/artist/${artistId}/check`);
            return {
                success: true,
                isFollowing: response.data.isFollowing || false
            };
        } catch (error) {
            console.error('Check following artist failed:', error);
            return {
                success: false,
                isFollowing: false,
                error: error.response?.data?.message || 'Failed to check follow status'
            };
        }
    },
};

export default followService;

File: aura-backend/src/main/java/com/aura/follow/service/FollowService.java
package com.aura.follow.service;

import com.aura.artist.entity.Artist;
import com.aura.artist.repository.ArtistRepository;
import com.aura.follow.dto.FollowDTO;
import com.aura.follow.dto.UserFollowStatsDTO;
import com.aura.follow.entity.Follow;
import com.aura.follow.repository.FollowRepository;
import com.aura.user.entity.User;
import com.aura.user.exception.DuplicateResourceException;
import com.aura.user.exception.ResourceNotFoundException;
import com.aura.user.repository.UserRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.function.Function;
import java.util.stream.Collectors;

@Service
public class FollowService {

    private final FollowRepository followRepository;
    private final UserRepository userRepository;
    private final ArtistRepository artistRepository;

    public FollowService(FollowRepository followRepository, UserRepository userRepository,
                        ArtistRepository artistRepository) {
        this.followRepository = followRepository;
        this.userRepository = userRepository;
        this.artistRepository = artistRepository;
    }

    private User resolveUserByEmail(String email) {
        if (email == null) return null;
        return userRepository.findByEmail(email)
            .orElseThrow(() -> new ResourceNotFoundException("User not found: " + email));
    }

    @Transactional
    public FollowDTO followUser(String followerEmail, Long followingId) {
        User follower = resolveUserByEmail(followerEmail);
        Long followerId = follower.getId();

        if (followerId.equals(followingId)) {
            throw new IllegalArgumentException("Cannot follow yourself");
        }

        User following = userRepository.findById(followingId)
            .orElseThrow(() -> new ResourceNotFoundException("User to follow not found"));

        if (followRepository.findByFollowerIdAndFollowingId(followerId, followingId).isPresent()) {
            throw new DuplicateResourceException("Already following this user");
        }

        Follow follow = new Follow();
        follow.setFollowerId(followerId);
        follow.setFollowingId(followingId);
        follow.setCreatedAt(LocalDateTime.now());

        Follow saved = followRepository.save(follow);
        return convertToDTO(saved, follower.getUsername(), following.getUsername());
    }

    @Transactional
    public void unfollowUser(String followerEmail, Long followingId) {
        User follower = resolveUserByEmail(followerEmail);
        Long followerId = follower.getId();

        Follow follow = followRepository.findByFollowerIdAndFollowingId(followerId, followingId)
            .orElseThrow(() -> new ResourceNotFoundException("Not following this user"));
        
        followRepository.delete(follow);
    }

    public boolean isFollowing(String followerEmail, Long followingId) {
        if (followerEmail == null) return false;
        User follower = resolveUserByEmail(followerEmail);
        return isFollowing(follower.getId(), followingId);
    }

    public boolean isFollowing(Long followerId, Long followingId) {
        if (followerId == null || followingId == null) return false;
        return followRepository.findByFollowerIdAndFollowingId(followerId, followingId).isPresent();
    }

    private UserFollowStatsDTO mapToStatsDTO(User user, Boolean isFollowing) {
        String displayName = user.getDisplayName() != null ? user.getDisplayName() : user.getUsername();
        String profileImageUrl = user.getProfileImageUrl() != null ? user.getProfileImageUrl() : user.getAvatarUrl();
        Boolean isArtist = false;
        
        Artist artist = artistRepository.findByUserId(user.getId());
        if (artist != null) {
            displayName = artist.getStageName();
            if (artist.getProfileImageUrl() != null) {
                profileImageUrl = artist.getProfileImageUrl();
            }
            isArtist = true;
        }
        
        return new UserFollowStatsDTO(
            user.getId(),
            user.getUsername(),
            followRepository.countByFollowingId(user.getId()),
            followRepository.countByFollowerId(user.getId()),
            isFollowing,
            displayName,
            profileImageUrl,
            isArtist
        );
    }

    private UserFollowStatsDTO mapArtistToStatsDTO(Artist artist, Boolean isFollowing) {
        return new UserFollowStatsDTO(
            artist.getArtistId(),
            artist.getStageName() != null ? artist.getStageName() : "Artist",
            followRepository.countByFollowingArtistId(artist.getArtistId()),
            0,
            isFollowing,
            artist.getStageName(),
            artist.getProfileImageUrl(),
            true
        );
    }

    public List<UserFollowStatsDTO> getFollowing(Long userId, String viewerEmail) {
        User viewer = viewerEmail != null ? userRepository.findByEmail(viewerEmail).orElse(null) : null;
        Long viewerId = viewer != null ? viewer.getId() : null;

        List<Follow> following = followRepository.findByFollowerId(userId);
        
        // Extract IDs for batch fetching
        List<Long> userIds = following.stream()
            .map(Follow::getFollowingId)
            .filter(Objects::nonNull)
            .collect(Collectors.toList());
            
        List<Long> artistIds = following.stream()
            .map(Follow::getFollowingArtistId)
            .filter(Objects::nonNull)
            .collect(Collectors.toList());

        // Batch fetch
        Map<Long, User> userMap = userRepository.findAllById(userIds).stream()
            .collect(Collectors.toMap(User::getId, Function.identity()));
            
        Map<Long, Artist> directArtistMap = artistRepository.findAllById(artistIds).stream()
            .collect(Collectors.toMap(Artist::getArtistId, Function.identity()));

        return following.stream().map(follow -> {
            if (follow.getFollowingId() != null) {
                User user = userMap.get(follow.getFollowingId());
                if (user != null) {
                    boolean isUserFollowing = isFollowing(viewerId, user.getId());
                    return mapToStatsDTO(user, isUserFollowing);
                }
            } else if (follow.getFollowingArtistId() != null) {
                Artist artist = directArtistMap.get(follow.getFollowingArtistId());
                if (artist != null) {
                    boolean isUserFollowingArtist = isFollowingArtist(viewerId, artist.getArtistId());
                    return mapArtistToStatsDTO(artist, isUserFollowingArtist);
                }
            }
            return null;
        }).filter(Objects::nonNull).collect(Collectors.toList());
    }

    public List<UserFollowStatsDTO> getFollowers(Long userId, String viewerEmail) {
        User viewer = viewerEmail != null ? userRepository.findByEmail(viewerEmail).orElse(null) : null;
        Long viewerId = viewer != null ? viewer.getId() : null;

        List<Follow> followers = followRepository.findByFollowingId(userId);
        
        // Extract IDs for batch fetching
        List<Long> userIds = followers.stream()
            .map(Follow::getFollowerId)
            .filter(Objects::nonNull)
            .collect(Collectors.toList());
            
        Map<Long, User> userMap = userRepository.findAllById(userIds).stream()
            .collect(Collectors.toMap(User::getId, Function.identity()));

        return followers.stream().map(follow -> {
            User user = userMap.get(follow.getFollowerId());
            if (user != null) {
                boolean isUserFollowing = isFollowing(viewerId, user.getId());
                return mapToStatsDTO(user, isUserFollowing);
            }
            return null;
        }).filter(Objects::nonNull).collect(Collectors.toList());
    }

    public UserFollowStatsDTO getUserStats(Long userId, String viewerEmail) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));
        
        Boolean isUserFollowing = isFollowing(viewerEmail, userId);
        return mapToStatsDTO(user, isUserFollowing);
    }

    public Integer getFollowersCount(Long userId) {
        return followRepository.countByFollowingId(userId);
    }

    public Integer getFollowingCount(Long userId) {
        return followRepository.countByFollowerId(userId);
    }

    // ================ ARTIST FOLLOWING ================

    @Transactional
    public FollowDTO followArtist(String followerEmail, Long artistId) {
        User follower = resolveUserByEmail(followerEmail);
        Long followerId = follower.getId();

        Artist artist = artistRepository.findById(artistId)
            .orElseThrow(() -> new ResourceNotFoundException("Artist not found"));
        
        if (artist.getUserId() != null) {
            return followUser(followerEmail, artist.getUserId());
        } else {
            if (followRepository.findByFollowerIdAndFollowingArtistId(followerId, artistId).isPresent()) {
                throw new DuplicateResourceException("Already following this artist");
            }
                
            Follow follow = new Follow();
            follow.setFollowerId(followerId);
            follow.setFollowingArtistId(artistId);
            follow.setCreatedAt(LocalDateTime.now());
            
            Follow saved = followRepository.save(follow);
            FollowDTO dto = convertToDTO(saved, follower.getUsername(), artist.getStageName());
            dto.setFollowingArtistId(artistId);
            return dto;
        }
    }

    @Transactional
    public void unfollowArtist(String followerEmail, Long artistId) {
        User follower = resolveUserByEmail(followerEmail);
        Long followerId = follower.getId();

        Artist artist = artistRepository.findById(artistId)
            .orElseThrow(() -> new ResourceNotFoundException("Artist not found"));
        
        if (artist.getUserId() != null) {
            unfollowUser(followerEmail, artist.getUserId());
        } else {
            Follow follow = followRepository.findByFollowerIdAndFollowingArtistId(followerId, artistId)
                .orElseThrow(() -> new ResourceNotFoundException("Not following this artist"));
            followRepository.delete(follow);
        }
    }

    public boolean isFollowingArtist(String followerEmail, Long artistId) {
        if (followerEmail == null) return false;
        User follower = resolveUserByEmail(followerEmail);
        return isFollowingArtist(follower.getId(), artistId);
    }

    public boolean isFollowingArtist(Long followerId, Long artistId) {
        if (followerId == null || artistId == null) return false;
        Artist artist = artistRepository.findById(artistId)
            .orElseThrow(() -> new ResourceNotFoundException("Artist not found"));
        
        if (artist.getUserId() != null) {
            return isFollowing(followerId, artist.getUserId());
        } else {
            return followRepository.findByFollowerIdAndFollowingArtistId(followerId, artistId).isPresent();
        }
    }

    private FollowDTO convertToDTO(Follow follow, String followerName, String followingName) {
        FollowDTO dto = new FollowDTO();
        dto.setFollowId(follow.getFollowId());
        dto.setFollowerId(follow.getFollowerId());
        dto.setFollowingId(follow.getFollowingId());
        dto.setFollowingArtistId(follow.getFollowingArtistId());
        dto.setFollowerName(followerName);
        dto.setFollowingName(followingName);
        dto.setCreatedAt(follow.getCreatedAt());
        return dto;
    }
}
